polkadot/node/network/statement-distribution/src/v2/mod.rs

```diff
@@ -3335,33 +3337,55 @@ pub(crate) async fn respond_task(
 	mut sender: mpsc::Sender<ResponderMessage>,
 ) {
 	let mut pending_out = FuturesUnordered::new();
+	let mut active_peers = HashSet::new();
+
 	loop {
-		// Ensure we are not handling too many requests in parallel.
-		if pending_out.len() >= MAX_PARALLEL_ATTESTED_CANDIDATE_REQUESTS as usize {
-			// Wait for one to finish:
-			pending_out.next().await;
-		}
+		select! {
+			// New request
+			request_result = receiver.recv(|| vec![COST_INVALID_REQUEST]).fuse() => {
+				let request = match request_result.into_nested() {
+					Ok(Ok(v)) => v,
+					Err(fatal) => {
+						gum::debug!(target: LOG_TARGET, error = ?fatal, "Shutting down request responder");
+						return
+					},
+					Ok(Err(jfyi)) => {
+						gum::debug!(target: LOG_TARGET, error = ?jfyi, "Decoding request failed");
+						continue
+					},
+				};
 
-		let req = match receiver.recv(|| vec![COST_INVALID_REQUEST]).await.into_nested() {
-			Ok(Ok(v)) => v,
-			Err(fatal) => {
-				gum::debug!(target: LOG_TARGET, error = ?fatal, "Shutting down request responder");
-				return
+				// If peer currently being served drop request
+				if active_peers.contains(&request.peer) {
```

Comment:
1. Are we 100 percent sure peers won't have more than one request in-flight ? 
2. What would a peer whose requests gets dropped see ?
3. If we don't expect this too happen, I would warn if two frequent, see warn_if_frequent.

Reply:
> Are we 100 percent sure peers won't have more than one request in-flight?

They can have more than 1 and this is the main reason why this change is getting implemented so we drop them when it happens. To reduce the wasted requester work there is also the rate limiter on the requester side so they don't request from the same peer.

Old nodes or malicious nodes can definitely send more requests. New nodes shouldn't have more than 1 in-flight, but even if something goes wrong and they have more than nothing should collapse.

> What would a peer whose requests gets dropped see?

First of all he shouldn't have sent it since he also should have known that he is requesting something else already. If it was an option he should have requested from a different peer that has the same data. 

If it happens anyway he would see the same thing as if the packet was lost along the way or if the packet was undecodable so AFAIK nothing. He might be potentially wasting our time so best to ignore it. Request will remain `in_flight` with no response and eventually will be reattempted `next_retry_time`. So the absolute worst case here as far as I can see is attempting again after the delay.

> If we don't expect this too happen, I would warn if two frequent, see warn_if_frequent.

We indeed don't expect this to happen ONCE people update. Old nodes can still send us too many requests since they will not have this rate limit so in this period this might happen more often that preferred.